# Segmenting and Clustering Neighborhoods in Toronto

### Hassan Ashiq 
- #### Connect with me on Linkdin : https://www.linkedin.com/in/hassan-ashiq

# *Part 1*

## Step 1 : Data Gathering from Wikipedia

In [218]:
import requests
from bs4 import BeautifulSoup
import pandas as pd 
import numpy as np
from sklearn.cluster import KMeans

In [99]:
page=requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
soup=BeautifulSoup(page.content, 'html.parser')


In [100]:
table_class='mw-parser-output'
items=soup.find(class_='mw-parser-output')
table_class=items.find(class_='wikitable sortable')
required_data=tabled.find_all('tr')

#### Creating DataFrame

In [137]:
#column_names = ['Postal-code','Borough', 'Neighborhood'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame()

In [138]:
for i in range(1,len(required_data)):
    td=required_data[i].find_all('td')
    postal_code=td[0].get_text().strip()
    borough=td[1].get_text().strip()
    neighbor=td[2].get_text().strip()
    
    neighborhoods=neighborhoods.append({'Postel-code':postal_code,
                         'Borough':borough,
                         'Neighborhood':neighbor}, ignore_index=True)

In [139]:
neighborhoods=neighborhoods[['Postel-code', 'Borough','Neighborhood']]

In [140]:
neighborhoods.shape

(180, 3)

## Step 2 Data Cleanup

#### 2.1 : removing rows with Not Assigned Borough

In [141]:
neighborhoods.drop(neighborhoods[neighborhoods.Borough=='Not assigned'].index, inplace=True)


#### 2.2 : More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.

In [142]:
neighborhoods["Neighborhood"] =neighborhoods.groupby("Postel-code")["Neighborhood"].transform(lambda neigh: ', '.join(neigh))

#### 2.3 : dropping duplicates

In [144]:
neighborhoods=neighborhoods.drop_duplicates()

#### 2.4 : If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough

In [147]:
neighborhoods['Neighborhood'].replace("Not assigned", neighborhoods["Borough"],inplace=True)

In [154]:
#updating the index 
neighborhoods=neighborhoods.reset_index(drop=True)

In [155]:
neighborhoods

Postel-code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

### Step 3 : Data Frame Shape Output

In [156]:
neighborhoods.shape

(103, 3)

# *Part 2*

### Step 4.a : Reading Geospatial Coordinates

In [168]:
coordinates_file=pd.read_csv('Geospatial_Coordinates.csv')
coordinates_file.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

*We will merge the two dataframes. To get the correct result independant on the order of data, we need to set indexes of two dataframes to its Postal Code columns*

In [170]:
neighborhoods.set_index('Postel-code', inplace=True)
coordinates_file.set_index('Postal Code', inplace=True)

*We will join the two files with respect to its Postal Codes*

In [175]:
toronto_df = pd.concat([neighborhoods, coordinates_file], axis=1, join='inner')

In [179]:
toronto_df.index.name="Postel-Code"

In [182]:
toronto_df.reset_index(inplace=True)

### Step 4.b : Output of DataFrame

In [183]:
toronto_df

Postel-Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

# *Part 3* : Explore and cluster the neighborhoods in Toronto

In [185]:
from geopy.geocoders import Nominatim
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans

In [188]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="tl-toronto-neigh")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.6534817, -79.3839347.


#### Please Note :
*You many not see the Folium Map in Github because GitHub doesnot support it* 

*You can view it in https://nbviewer.jupyter.org/*

*just put this notebook link there.*

#### Using Folium to create a Map of Toronto with Boroughs markers on top

In [196]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, long, post, borough, neigh in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Postel-Code'], toronto_df['Borough'], toronto_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

*Reduce the number of Boroughs to explore To reduce the numbers of calls to FourSquare API, we will *only explore boroughs that have Toronto in their names.*

In [193]:
toronto_boroughs = ['East Toronto', 'Central Toronto', 'Downtown Toronto', 'West Toronto']
toronto_central_df = toronto_df[toronto_df['Borough'].isin(toronto_boroughs)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()

(39, 5)


Postel-Code           Borough                                 Neighborhood  \
0         M5A  Downtown Toronto                    Regent Park, Harbourfront   
1         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2         M5B  Downtown Toronto                     Garden District, Ryerson   
3         M5C  Downtown Toronto                               St. James Town   
4         M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031

In [197]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=12)

for lat, long, post, borough, neigh in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['Postel-Code'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    label = "{} ({}): {}".format(borough, post, neigh)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

#### *Using FourSquare API to explore the Boroughs*

In [203]:
CLIENT_ID = 'MQQXHBN1NLGAYJG445WR5EWZ0OSI4SP2YKWMHSU53AQB3MF2' # your Foursquare ID
CLIENT_SECRET = 'GKQC1M3LQFGRCGBHWQABJA1CG0CLPP0FDK1UG4X4YFZV2ANK' # your Foursquare Secret
VERSION = '20190330' # Foursquare API version
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: MQQXHBN1NLGAYJG445WR5EWZ0OSI4SP2YKWMHSU53AQB3MF2
CLIENT_SECRET:GKQC1M3LQFGRCGBHWQABJA1CG0CLPP0FDK1UG4X4YFZV2ANK


In [204]:
radius = 500
LIMIT = 100

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['Postel-Code'], toronto_central_df['Borough'], toronto_central_df['Neighborhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [222]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['Postel-Code', 'Borough', 'Neighborhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.shape)
venues_df.head()

(1620, 9)


Postel-Code           Borough               Neighborhood  BoroughLatitude  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
1         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
2         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
3         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   
4         M5A  Downtown Toronto  Regent Park, Harbourfront         43.65426   

   BoroughLongitude                 VenueName  VenueLatitude  VenueLongitude  \
0        -79.360636          Roselle Desserts      43.653447      -79.362017   
1        -79.360636             Tandem Coffee      43.653559      -79.361809   
2        -79.360636    Cooper Koo Family YMCA      43.653249      -79.358008   
3        -79.360636       Body Blitz Spa East      43.654735      -79.359874   
4        -79.360636  Dominion Pub and Kitchen      43.656919      -79.358967   

         VenueCategory  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4                  Pub

In [223]:
venues_df.groupby(['Postel-Code', 'Borough', 'Neighborhood'])['VenueName'].count()

Postel-Code  Borough           Neighborhood                                                                                              
M4E          East Toronto      The Beaches                                                                                                     5
M4K          East Toronto      The Danforth West, Riverdale                                                                                   43
M4L          East Toronto      India Bazaar, The Beaches West                                                                                 24
M4M          East Toronto      Studio District                                                                                                40
M4N          Central Toronto   Lawrence Park                                                                                                   4
M4P          Central Toronto   Davisville North                                                                                          

In [224]:
len(venues_df['VenueCategory'].unique())

237

so there are 237 unique venues

#### Analyze venues in each area

In [225]:
# one hot encoding
toronto_central_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_central_onehot['Postel-Code'] = venues_df['Postel-Code'] 
toronto_central_onehot['Borough'] = venues_df['Borough'] 
toronto_central_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_central_onehot.columns[-3:]) + list(toronto_central_onehot.columns[:-3])
toronto_central_onehot = toronto_central_onehot[fixed_columns]

print(toronto_central_onehot.shape)
toronto_central_onehot.head()

(1620, 240)


Postel-Code           Borough              Neighborhoods  Afghan Restaurant  \
0         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
1         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
2         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
3         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   
4         M5A  Downtown Toronto  Regent Park, Harbourfront                  0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0        0                   0             0               0                0   
1        0                   0             0               0                0   
2        0                   0             0               0                0   
3        0                   0             0               0                0   
4        0                   0             0               0                0   

   Airport Terminal  ...  Toy / Game Store  Trail  Train Station  \
0                 0  ...                 0      0              0   
1                 0  ...                 0      0              0   
2                 0  ...                 0      0              0   
3                 0  ...                 0      0              0   
4                 0  ...                 0      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0         0          0              0            0  
1         0          0              0            0  
2         0          0              0            0  
3         0          0              0            0  
4         0          0              0            0  

[5 rows x 240 columns]

#### Get the frequency of occurance of each category in an area

In [226]:
toronto_central_venues_freq = toronto_central_onehot.groupby(['Postel-Code', 'Borough', 'Neighborhoods']).mean().reset_index()
print(toronto_central_venues_freq.shape)
toronto_central_venues_freq.head()

(39, 240)


Postel-Code          Borough                   Neighborhoods  \
0         M4E     East Toronto                     The Beaches   
1         M4K     East Toronto    The Danforth West, Riverdale   
2         M4L     East Toronto  India Bazaar, The Beaches West   
3         M4M     East Toronto                 Studio District   
4         M4N  Central Toronto                   Lawrence Park   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  ...  Toy / Game Store  \
0             0.0              0.0               0.0  ...               0.0   
1             0.0              0.0               0.0  ...               0.0   
2             0.0              0.0               0.0  ...               0.0   
3             0.0              0.0               0.0  ...               0.0   
4             0.0              0.0               0.0  ...               0.0   

      Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0  0.200000            0.0                            0.0               0.0   
1  0.023256            0.0                            0.0               0.0   
2  0.000000            0.0                            0.0               0.0   
3  0.000000            0.0                            0.0               0.0   
4  0.000000            0.0                            0.0               0.0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0                    0.0     0.000        0.0            0.0     0.000000  
1                    0.0     0.000        0.0            0.0     0.023256  
2                    0.0     0.000        0.0            0.0     0.000000  
3                    0.0     0.025        0.0            0.0     0.025000  
4                    0.0     0.000        0.0            0.0     0.000000  

[5 rows x 240 columns]

#### Get 10 most occurance venue types in each area

In [228]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['Postel-Code', 'Borough', 'Neighborhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns
# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Postel-Code'] = toronto_central_venues_freq['Postel-Code']
neighborhoods_venues_sorted['Borough'] = toronto_central_venues_freq['Borough']
neighborhoods_venues_sorted['Neighborhoods'] = toronto_central_venues_freq['Neighborhoods']

for ind in np.arange(toronto_central_venues_freq.shape[0]):
    row_categories = toronto_central_venues_freq.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighborhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

neighborhoods_venues_sorted.sort_values(freqColumns, inplace=True)
neighborhoods_venues_sorted

Postel-Code           Borough  \
27         M5V  Downtown Toronto   
31         M6H      West Toronto   
32         M6J      West Toronto   
33         M6K      West Toronto   
3          M4M      East Toronto   
15         M5C  Downtown Toronto   
25         M5S  Downtown Toronto   
34         M6P      West Toronto   
24         M5R   Central Toronto   
26         M5T  Downtown Toronto   
14         M5B  Downtown Toronto   
6          M4R   Central Toronto   
19         M5J  Downtown Toronto   
11         M4X  Downtown Toronto   
28         M5W  Downtown Toronto   
36         M6S      West Toronto   
18         M5H  Downtown Toronto   
21         M5L  Downtown Toronto   
29         M5X  Downtown Toronto   
16         M5E  Downtown Toronto   
20         M5K  Downtown Toronto   
13         M5A  Downtown Toronto   
17         M5G  Downtown Toronto   
37         M7A  Downtown Toronto   
12         M4Y  Downtown Toronto   
38         M7Y      East Toronto   
22         M5N   Central Toronto   
35         M6R      West Toronto   
1          M4K      East Toronto   
30         M6G  Downtown Toronto   
0          M4E      East Toronto   
4          M4N   Central Toronto   
2          M4L      East Toronto   
5          M4P   Central Toronto   
10         M4W  Downtown Toronto   
8          M4T   Central Toronto   
7          M4S   Central Toronto   
9          M4V   Central Toronto   
23         M5P   Central Toronto   

                                        Neighborhoods 1st Most Common Venue  \
27  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
31                       Dufferin, Dovercourt Village                Bakery   
32                           Little Portugal, Trinity                   Bar   
33       Brockton, Parkdale Village, Exhibition Place                  Café   
3                                     Studio District                  Café   
15                                     St. James Town                  Café   
25                     University of Toronto, Harbord                  Café   
34                      High Park, The Junction South                  Café   
24                The Annex, North Midtown, Yorkville                  Café   
26          Kensington Market, Chinatown, Grange Park                  Café   
14                           Garden District, Ryerson        Clothing Store   
6                  North Toronto West,  Lawrence Park        Clothing Store   
19  Harbourfront East, Union Station, Toronto Islands           Coffee Shop   
11                        St. James Town, Cabbagetown           Coffee Shop   
28                                     Stn A PO Boxes           Coffee Shop   
36                                 Runnymede, Swansea           Coffee Shop   
18                           Richmond, Adelaide, King           Coffee Shop   
21                     Commerce Court, Victoria Hotel           Coffee Shop   
29             First Canadian Place, Underground city           Coffee Shop   
16                                        Berczy Park           Coffee Shop   
20           Toronto Dominion Centre, Design Exchange           Coffee Shop   
13                          Regent Park, Harbourfront           Coffee Shop   
17                                 Central Bay Street           Coffee Shop   
37        Queen's Park, Ontario Provincial Government           Coffee Shop   
12                               Church and Wellesley           Coffee Shop   
38  Business reply mail Processing Centre, South C...                Garden   
22                                           Roselawn                Garden   
35                             Parkdale, Roncesvalles             Gift Shop   
1                        The Danforth West, Riverdale      Greek Restaurant   
30                                           Christie         Grocery Store   
0                                         The Beaches     Health Food Store   
4                                 

#### Clustering areas

In [230]:

kclusters = 4

toronto_central_venues_freq_clustering = toronto_central_venues_freq.drop(['Postel-Code', 'Borough', 'Neighborhoods'], 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_central_venues_freq_clustering)

toronto_central_clustered_df = toronto_central_df

toronto_central_clustered_df['Cluster'] = kmeans.labels_

toronto_central_clustered_df = toronto_central_clustered_df.join(neighborhoods_venues_sorted.drop(['Borough', 'Neighborhoods'], 1).set_index('Postel-Code'), on='Postel-Code')
toronto_central_clustered_df.sort_values(['Cluster'] + freqColumns, inplace=True)
toronto_central_clustered_df


Postel-Code           Borough  \
23         M4R   Central Toronto   
10         M5J  Downtown Toronto   
8          M5H  Downtown Toronto   
32         M5V  Downtown Toronto   
9          M6H      West Toronto   
11         M6J      West Toronto   
14         M6K      West Toronto   
17         M4M      East Toronto   
3          M5C  Downtown Toronto   
27         M5S  Downtown Toronto   
24         M5R   Central Toronto   
30         M5T  Downtown Toronto   
2          M5B  Downtown Toronto   
35         M4X  Downtown Toronto   
34         M5W  Downtown Toronto   
28         M6S      West Toronto   
16         M5L  Downtown Toronto   
36         M5X  Downtown Toronto   
5          M5E  Downtown Toronto   
13         M5K  Downtown Toronto   
0          M5A  Downtown Toronto   
6          M5G  Downtown Toronto   
1          M7A  Downtown Toronto   
37         M4Y  Downtown Toronto   
38         M7Y      East Toronto   
19         M5N   Central Toronto   
25         M6R      West Toronto   
12         M4K      East Toronto   
7          M6G  Downtown Toronto   
18         M4N   Central Toronto   
15         M4L      East Toronto   
20         M4P   Central Toronto   
33         M4W  Downtown Toronto   
29         M4T   Central Toronto   
26         M4S   Central Toronto   
31         M4V   Central Toronto   
21         M5P   Central Toronto   
22         M6P      West Toronto   
4          M4E      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
23                 North Toronto West,  Lawrence Park  43.715383 -79.405678   
10  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752   
8                            Richmond, Adelaide, King  43.650571 -79.384568   
32  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420   
9                        Dufferin, Dovercourt Village  43.669005 -79.442259   
11                           Little Portugal, Trinity  43.647927 -79.419750   
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191   
17                                    Studio District  43.659526 -79.340923   
3                                      St. James Town  43.651494 -79.375418   
27                     University of Toronto, Harbord  43.662696 -79.400049   
24                The Annex, North Midtown, Yorkville  43.672710 -79.405678   
30          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049   
2                            Garden District, Ryerson  43.657162 -79.378937   
35                        St. James Town, Cabbagetown  43.667967 -79.367675   
34                                     Stn A PO Boxes  43.646435 -79.374846   
28                                 Runnymede, Swansea  43.651571 -79.484450   
16                     Commerce Court, Victoria Hotel  43.648198 -79.379817   
36             First Canadian Place, Underground city  43.648429 -79.382280   
5                                         Berczy Park  43.644771 -79.373306   
13           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576   
0                           Regent Park, Harbourfront  43.654260 -79.360636   
6                                  Central Bay Street  43.657952 -79.387383   
1         Queen's Park, Ontario Provincial Government  43.662301 -79.389494   
37                               Church and Wellesley  43.665860 -79.383160   
38  Business reply mail Processing Centre, South C...  43.662744 -79.321558   
19                                           Roselawn  43.711695 -79.416936   
25                             Parkdale, Roncesvalles  43.648960 -79.456325   
12                       The Danforth West, Riverdale  43.679557 -79.352188   
7                                            Christie  43.669542 -79.422564   
18                                      Lawrence Park  43.728020 -79.388790   
15                     India Bazaar, The Beaches West  43.668999 -79.315572   
20                                

#### Show those clusters onto a map


In [231]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster in zip(toronto_central_clustered_df['Latitude'], toronto_central_clustered_df['Longitude'], toronto_central_clustered_df['Postel-Code'], toronto_central_clustered_df['Borough'], toronto_central_clustered_df['Neighborhood'], toronto_central_clustered_df['Cluster']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters


**Upon observing the result, we can name the clusters as follow:**

- *Cluster 0 (Red): Living area (with mostly park, trail, school; and some small businesses)*

- *Cluster 1 (Yellow): Roselawn - Central Toronto (nothing here except a garden)*

- *Cluster 2 (Purple): Business area (with lots of business venues)*

- *Cluster 3 (Blue) : Oldest neighboorhood with lot of nature park*

**Thank for following my project.**

### Hassan Ashiq 
- #### Connect with me on Linkdin : https://www.linkedin.com/in/hassan-ashiq